hal/src/dma.rs

```diff
@@ -0,0 +1,1075 @@
+
+            // Initialize descriptor list addresses
+            self.dmac.baseaddr.write(|reg| unsafe {
+                reg.baseaddr()
+                    .bits(self.descriptor_list.descriptor.as_ptr() as u32)
+            });
+
+            self.dmac.wrbaddr.write(|reg| unsafe {
+                reg.wrbaddr()
+                    .bits(self.descriptor_list.writeback.as_ptr() as u32)
+            });
+
+            // TODO(tarcieri): Wipe decriptor and writeback tables
+            // self.descriptor_list.descriptor[0] = Descriptor::default();
+            // self.descriptor_list.descriptor[0] = Descriptor::default();
+
+            // Re-enable DMA controller with all priority levels
+            self.dmac.ctrl.write(|reg| {
+                reg.dmaenable()
+                    .set_bit()
+                    .lvlen0()
+                    .set_bit()
+                    .lvlen1()
+                    .set_bit()
+                    .lvlen2()
+                    .set_bit()
+                    .lvlen3()
+                    .set_bit()
+            });
+
+            // Enable DMA interrupt at lowest priority
+            #[cfg(feature = "samd51")]
+            {
+                // Enable DMAC IRQs
+                for irq in &[
+                    IRQn::N0,
+                    IRQn::N1,
+                    IRQn::N2,
+                    IRQn::N3,
+                    IRQn::N4,
+                ] {
+                    nvic_enable_irq(*irq);
+                    nvic_set_priority(*irq, (1 << NVIC_PRIO_BITS) - 1);
+                }
+            }
+
+            #[cfg(not(feature = "samd51"))]
+            {
+                // Enable DMAC IRQ
+                nvic_enable_irq(IRQN);
+                nvic_set_priority(IRQN, (1 << NVIC_PRIO_BITS) - 1);
+            }
+        }
+
+        self.channel = Some(channel);
+        int_state
+            .channel_mask
+            .set(int_state.channel_mask.get() | 1 << channel); // Mark channel as allocated
+
+        // Reset the allocated channel
+        #[cfg(feature = "samd51")]
+        {
+            self.dmac
+                .chctrla(channel)
+                .write(|reg| reg.enable().clear_bit().swrst().set_bit());
+        }
+
+        #[cfg(not(feature = "samd51"))]
+        unsafe {
+            self.dmac.chid.write(|reg| reg.id().bits(channel));
+            self.dmac
+                .chctrla
+                .write(|reg| reg.enable().clear_bit().swrst().set_bit());
+        }
+
+        // Clear software trigger
+        self.dmac
+            .swtrigctrl
+            .modify(|reg, value| unsafe { value.bits(reg.bits() & !(1 << channel)) });
+
+        // This type unfortunately does not impl Copy or Clone
+        let peripheral_trigger = match self.peripheral_trigger {
+            TRIGSRCW::DISABLE => TRIGSRCW::DISABLE,
+        };
+
+        // Ditto for this type
+        let trigger_action = match self.trigger_action {
```

Comment:
These are annoying and ugly as `TRIGSRCW` and `TRIGACTW` don't impl `Copy` or `Clone`. Any ideas how to work around that? A newtype maybe?